### PR TITLE
feat(wallet): implement solana dapps send tx

### PR DESCRIPTION
**Pull request by Pavneet-Sing**
- fix show wallet icon with sol dapps request

Resolves https://github.com/brave/brave-browser/issues/25419
Resolves https://github.com/brave/brave-browser/issues/25258

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

[solna-dapps-tx.webm](https://user-images.githubusercontent.com/11755381/197188431-5fcedac5-1f38-44b4-8827-9248f604143d.webm)

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:441]: `TextView tvLabel = AndroidUtils.makeHeaderTv(requireContext());
  TextView tvVal = AndroidUtils.makeSubHeaderTv(requireContext());`
could be moved out of 3 ifs as they are same and called once before the conditions

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:445]: the same with
```
                    sendOptionsLl.addView(tvLabel);
                    sendOptionsLl.addView(tvVal);
```
but should be called once after all conditions.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java:65]: I don't get why do we need to call `safeInit` here if `mValue` and `value` are unrelated. That function has to static perhaps as well

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java:70]: same here as above, how `Amount value` and that whole class are related? Why do we need to call some `safeInit` of that class if we don't use `mValue` from it at all?

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java:153]: couldn't `mSolanaInstruction` be null? Shouldn't we do a check on that? I think it can as a default constructor `SolanaInstructionPresenter` is accessible.

Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java:153]: the above is related to all usages of `mSolanaInstruction` thought the class

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:52]: why do we call the function as general `makeHeaderTv` if it's related to solana only. `AndroidUtils.java` contains general utils helper functions for the whole crypto wallet. We should call it somehow `makeSolanaHeaderTv` if it's relate to Solana only or pass the string as an argument, so it could be used everywhere.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:61]: the same as above regarding a function name.

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:441]: They all are three different textviews not one, just using the generic name. We only show these if a corresponding value exists.

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:445]: Same as [this](https://github.com/brave/brave-core/pull/15593#discussion_r1007700384), all three are different.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java:65]: It is just an extra check to make sure the existing value is never null to avoid any potential exception. I made `mValue` private for time being but may change it to `public`.
We shouldn't/can't use static here because `mValue` is not static.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java:70]: here it is not required. will remove.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java:153]: I have an assert here for [null check](https://github.com/brave/brave-core/blob/b10d98c011cb6640447589b1a143b90cabd04732/android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java#L39).
There is no default constructor because java doesn't provide one if we define a parameterised constructor.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:52]: It's just a general function to create a header textview with a bold style. Don't wanna pass a string as it could be null etc. so don't wanna make it more specific than creating a header textview. we can make more helper function in the future if it is required.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:61]: Same as above, just a normal textview to be used as a sub-header with no additional checks etc to avoid extra responsibilities or checks. Can add more helper later if required.

**Comment by Pavneet-Sing** [comment]
Skipping other tests expect android as only making android changes now.

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:441]: could they be shown alltogether in some cases?

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:52]: so why do we do `textView.setText(R.string.brave_wallet_solana_max_retries);` if it's just to create a header textview, why not some other string. We should remove `setText` if it's just a helper function to create a header textview with a bold style or move it out of utils if it's only to show that particular string and we cannot reuse it for other strings

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:61]: same question as above, why `textView.setText(R.string.brave_wallet_solana_max_retries);` and not `textView.setText(R.string.ok);`?

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:441]: yes, as per the current test case and it's safe to keep the scope limited per case.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: we don't use `mValue` in `private BigDecimal safeVal(BigDecimal value) {` at all. That function doesn't access or doesn't modify anything from `Amount` class. Why it cannot be static?

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:52]: missed to remove that `setText` call, causing the confusion, will remove.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: same as above I don't get why it's not static. Any class methods have to manipulate with class members or access to them for a read only purpose. But a class function that makes a decision based on a passed value shouldn't have an access to the class if it doesn't need it.

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/AndroidUtils.java:61]: as above, missed to remove that, will remove, thanks!

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: Practically, we can make any function static which does not use member function, but the amount class is not a helper Utility class so better to avoid static functions separate in a Utility class.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: Same as [above](https://github.com/brave/brave-core/pull/15593#discussion_r1007970736).

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java
Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: So tell me please, why Amount class has to be created to do that?
```
    private BigDecimal safeVal(BigDecimal value) {
        return (value == null) ? BigDecimal.ZERO : value;
    }
```
why do we need this pointer accessible to some unrelated class at all to that function.
I will be ok to be it like that if you convince me that Amount class creation is necessary for that function execution.

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: If someone requires to use a particular method of a class then it can be converted to static though pragmatically, it can cause issues like:
- We won't be able to use other methods of amount class if we are directly using the safeVal
- It's a private method which can be changed anytime so shouldn't be directly accessible
I hope it make sense.

Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: no it's not, I can say for any helper function in Utils that I need it to be inside some particular class as it can be changed anytime. I don't see anything why that class is `private` and cannot be used outside of that class and why it's not static and keeps a reference on `this` pointer on some unrelated class.

Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: and could you please give an answer to that question `why Amount class has to be created to do that?`

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: 1. Pragmatically, we shouldn't make a method public if we are not sure or it can be modified.
2. As I have said earlier, can make it static if needed but better to keep it in a Utils class of static method to keep similar things together,

`why Amount class has to be created to do that?`
Usage of any class is not forced but the purpose of a class is to offer similar features.

Let's ask the opinions from others: @samartnik @AlexeyBarabash @deeppandya

Comment (SergeyZhukovsky) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: use `private static` for such functions and it won't be accessible directly from other places

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: updated https://github.com/brave/brave-core/blob/54e2cbc2217f1a43e1340bafd9a39b869dac6e6b/android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java#L64

Comment (AlexeyBarabash) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/Amount.java]: @Pavneet-Sing 
I would also use `static` for  `safeVal` as it doesn't use any members. Basically, when I see the static method in Java or C++, I instantly understand that the method doesn't use access to members. 

About to put or not into Utils class - in my opinion, static methods often are by itself something like class-internals-util method. So private static seems good.

If `safeVal(BigDecimal value)` will be required to be used outside of `Amount` then move it to `Utils`.

File: android/java/res/values/brave_styles.xml
Comment (AlexeyBarabash) [inline android/java/res/values/brave_styles.xml:189]: :+1: for new themes separated fro Chromium's

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxDetailsFragment.java
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxDetailsFragment.java:63]: thanks for this simplification

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:422]: nit: `sendOptionsLl` is hard to read, first I understood it as sendOptionsL_one and then tried to find `sendOptionsL2`. 
then I understood that `Ll` is for LinearLayout type. 
So, since there is no any other layouts here for `sendOptions`,  I would just rename it 
 `sendOptionsLl` => ` `sendOptionsLayout`
 wdyt?

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java:422]: That is short for LinerarLayout i.e. `Ll` (though look like `L1`), to give an indication that it's a LinearLayout, will update.

File: android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java:195]: the part of  `fromPubKey()`  which extracts pubkey is the same as one at `toPubKey()`
the only difference is `WalletConstants.SOL_DAPP_FROM_ACCOUNT` vs `WalletConstants.SOL_DAPP_TO_ACCOUNT`

Is it possible to make something like `String getMetaPubKeyByParam(String paramName)` and use it on both `fromPubKey()` and `toPubKey()` ?

Comment (AlexeyBarabash) [reply android/java/org/chromium/chrome/browser/crypto_wallet/presenters/SolanaInstructionPresenter.java:195]: oh, `getPubKeyPerParamKey` below does something similar, so can it also   re-use the most general method?

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransaction.java
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransaction.java:143]: ++ for direct use of `txDataUnion`

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransaction.java
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransaction.java:265]: :) thanks for cleaning this but the previous one on line 149 slipped through

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransaction.java:265]: That is intentional, it is just an extra unwanted semi-colon, so removed it :)

Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransaction.java:265]: will remove it :) thanks

File: android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransactionFees.java
Comment (AlexeyBarabash) [inline android/java/org/chromium/chrome/browser/crypto_wallet/util/ParsedTransactionFees.java:109]: :+1:

**Review by AlexeyBarabash — APPROVED** [review]
LGTM but please wait for approval from guys who recently worked with Android wallet

File: android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java
Comment (Pavneet-Sing) [reply android/java/org/chromium/chrome/browser/crypto_wallet/fragments/TxFragment.java]: updated: https://github.com/brave/brave-core/pull/15593/commits/11b346d25a205d33ccba689d0f40a713ee019466

**Comment by Pavneet-Sing** [comment]
> LGTM but please wait for approval from guys who recently worked with Android wallet

That would be @SergeyZhukovsky so it's good to merge.

**Comment by srirambv** [comment]
Verification passed on Oppo Reno 5 with Android 12 running 1.47.42 x64 Nightly build
- Verified steps from PR description
- Verified wallet icon is shown on URL bar when a Dapp is connected
- Verified able to submit and sign transactions in Solana Dapps

https://user-images.githubusercontent.com/17010094/199972976-b97ced61-495f-4bfd-8f15-ce4e8b87a8d8.mp4

**Pull request merged**
